File: frontend/src/services/alert.ts
import { api } from "./api";

export type AlertType = "PRICE_ABOVE" | "PRICE_BELOW" | "CHANGE_PCT_UP" | "CHANGE_PCT_DOWN";

export interface AlertOut {
  id: number;
  asset_id: number;
  asset_name: string;
  asset_symbol: string;
  alert_type: AlertType;
  threshold: number;
  is_active: boolean;
  last_triggered_at: string | null;
  created_at: string;
}

export interface AlertCreate {
  asset_id: number;
  alert_type: AlertType;
  threshold: number;
}

export interface AlertUpdate {
  is_active?: boolean;
  threshold?: number;
}

export const alertService = {
  async list(): Promise<AlertOut[]> {
    const { data } = await api.get<AlertOut[]>("/alerts");
    return data;
  },

  async create(body: AlertCreate): Promise<AlertOut> {
    const { data } = await api.post<AlertOut>("/alerts", body);
    return data;
  },

  async update(id: number, body: AlertUpdate): Promise<AlertOut> {
    const { data } = await api.patch<AlertOut>(`/alerts/${id}`, body);
    return data;
  },

  async remove(id: number): Promise<void> {
    await api.delete(`/alerts/${id}`);
  },
};

File: backend/app/schemas/alert.py
import datetime

from pydantic import BaseModel, field_validator

from app.models.alert import AlertType


class AlertCreate(BaseModel):
    asset_id: int
    alert_type: AlertType
    threshold: float

    @field_validator("threshold")
    @classmethod
    def threshold_positive(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("La soglia deve essere positiva")
        return v


class AlertUpdate(BaseModel):
    is_active: bool | None = None
    threshold: float | None = None


class AlertOut(BaseModel):
    id: int
    asset_id: int
    asset_name: str
    asset_symbol: str
    alert_type: str
    threshold: float
    is_active: bool
    last_triggered_at: datetime.datetime | None
    created_at: datetime.datetime
